src/Storages/Hive/HiveCommon.cpp

```diff
@@ -0,0 +1,243 @@
+HiveMetastoreClient::getTableMetadata(const String & db_name, const String & table_name)
+{
+    LOG_TRACE(log, "Get table metadata for {}.{}", db_name, table_name);
+    std::lock_guard lock{mutex};
+
+    auto table = std::make_shared<Apache::Hadoop::Hive::Table>();
+    std::vector<Apache::Hadoop::Hive::Partition> partitions;
+    try
+    {
+        client->get_table(*table, db_name, table_name);
+
+        /// Query the latest partition info to check new change.
+        client->get_partitions(partitions, db_name, table_name, -1);
+    }
+    catch (apache::thrift::transport::TTransportException & e)
+    {
+        setExpired();
+        throw Exception(ErrorCodes::NO_HIVEMETASTORE, "Hive Metastore expired because {}", String(e.what()));
+    }
+
+    bool update_cache = shouldUpdateTableMetadata(db_name, table_name, partitions);
+    String cache_key = db_name + "." + table_name;
+
+    std::shared_ptr<HiveMetastoreClient::HiveTableMetadata> metadata = table_metadata_cache.get(cache_key);
+
+    if (update_cache)
+    {
+        LOG_INFO(log, "Reload hive partition metadata info for {}.{}", db_name, table_name);
+
+        /// Generate partition infos from partitions and old partition infos(if exists).
+        std::map<String, PartitionInfo> new_partition_infos;
+        if (metadata)
+        {
+            auto & old_partiton_infos = metadata->getPartitionInfos();
+            for (const auto & partition : partitions)
+            {
+                auto it = old_partiton_infos.find(partition.sd.location);
+                if (it == old_partiton_infos.end() || !it->second.haveSameParameters(partition) || !it->second.initialized)
+                {
+                    new_partition_infos.emplace(partition.sd.location, PartitionInfo(partition));
+                    continue;
+                }
+                else
+                {
+                    PartitionInfo new_partition_info(partition);
+                    new_partition_info.files = std::move(it->second.files);
+                    new_partition_info.initialized = true;
+                }
+            }
+        }
+        else
+        {
+            for (const auto & partition : partitions)
+                new_partition_infos.emplace(partition.sd.location, PartitionInfo(partition));
+        }
+
+        metadata = std::make_shared<HiveMetastoreClient::HiveTableMetadata>(
+            db_name, table_name, table, std::move(new_partition_infos), getContext());
+        table_metadata_cache.set(cache_key, metadata);
+    }
+    return metadata;
+}
+
+void HiveMetastoreClient::clearTableMetadata(const String & db_name, const String & table_name)
+{
+    String cache_key = getCacheKey(db_name, table_name);
+
+    std::lock_guard lock{mutex};
+    std::shared_ptr<HiveMetastoreClient::HiveTableMetadata> metadata = table_metadata_cache.get(cache_key);
+    if (metadata)
+        table_metadata_cache.set(cache_key, nullptr);
```

Comment:
Why not remove the key completely?

Reply:
LRUCache doesn't have a del() method. And again, To add this method seems to make the evict policy more and more complicated in LRUCache

Reply:
I am in doubt about `set(key, nullptr)` idea as there are places where you call for example`canRelease(*cell.value)` without checking for nullptr.